# apply_to: 25.*.*
diff --git a/dlls/ntdll/unix/loader.c b/dlls/ntdll/unix/loader.c
index 59ebc37..6483140 100644
--- a/dlls/ntdll/unix/loader.c
+++ b/dlls/ntdll/unix/loader.c
@@ -499,7 +499,8 @@ static void init_paths( char *argv[] )
      * This isn't usable for building the path to the wineloader, so we need to hardcode
      * the binary's name (which is just "wine" in upstream, but CrossOver hacks it to be "wineloader").
      */
-    basename = "wineloader";
+    /* AZ HACK: revert to "wine" for CXBuilder */
+    basename = "wine";
 #endif
 
     if (!dladdr( init_paths, &info ) || !(ntdll_dir = realpath_dirname( info.dli_fname )))
